packages/next/src/lib/metadata/resolvers/resolve-url.ts

```diff
@@ -93,10 +93,33 @@ function resolveAbsoluteUrlWithPathname(
     canonicalUrl = result.pathname === '/' ? result.origin : result.href
   }
 
-  // Add trailing slash if it's enabled
-  return trailingSlash && !canonicalUrl.endsWith('/')
-    ? `${canonicalUrl}/`
-    : canonicalUrl
+  let isRelative = canonicalUrl.startsWith('/')
+  let isExternal = false
+  let hasQuery = canonicalUrl.includes('?')
+  if (!isRelative) {
+    try {
+      const parsedUrl = new URL(canonicalUrl)
```

Comment:
Seems like we could add these flags when we are constructing the canonicalURL rather than re-creating a URL object.  Something like:

```tsx
  // Get canonicalUrl without trailing slash
  let canonicalUrl = ''
  let isExternal = false
  let hasQuery = false
  const result = metadataBase ? resolveUrl(url, metadataBase) : url
  if (typeof result === 'string') {
    canonicalUrl = result
    hasQuery = canonicalUrl.includes('?')
  } else {
    canonicalUrl = result.pathname === '/' ? result.origin : result.href
    isExternal = result.origin !== metadataBase?.origin
    hasQuery = result.search !== ''
  }
  ```

Reply:
even it's string, we need need to know if it's the same origin url and append trialing slash if possible